pkg/pgmodel/sql_ingest.go

```diff
@@ -539,38 +542,115 @@ func runInserter(conn pgxConn, in chan copyRequest) {
 	// We grab copyRequests off the channel one at a time. This, and the name is
 	// a legacy from when we used CopyFrom to perform the insetions, and may
 	// change in the future.
+	insertBatch := make([]copyRequest, 0, maxCopyRequestsPerTxn)
 	for {
-		req, ok := <-in
+		var ok bool
+		insertBatch, ok = inserterGetBatch(insertBatch, in)
 		if !ok {
 			return
 		}
-		err := doInsert(conn, req)
+
+		// sort to prevent deadlocks
+		sort.Slice(insertBatch, func(i, j int) bool {
+			return insertBatch[i].table < insertBatch[j].table
+		})
+
+		//merge tables to prevent deadlocks (on row order) and for efficiency
+		for i := 0; i < len(insertBatch); i++ {
+			//for loop since may need to merge multiple table
+			for i < len(insertBatch)-2 && insertBatch[i].table == insertBatch[i+1].table {
+				insertBatch[i].data.absorb(insertBatch[i+1].data)
+				insertBatch[i+1].data.release()
+				//delete element i+1
+				copy(insertBatch[i+1:], insertBatch[i+2:])
+				insertBatch[len(insertBatch)-1] = copyRequest{}
+				insertBatch = insertBatch[:len(insertBatch)-1]
+			}
+		}
+
+		doInsertOrFallback(conn, insertBatch...)
+		for i := range insertBatch {
+			insertBatch[i] = copyRequest{}
+		}
+		insertBatch = insertBatch[:0]
+	}
+}
+
+var getBatchMutex = &sync.Mutex{}
+
+func inserterGetBatch(batch []copyRequest, in chan copyRequest) ([]copyRequest, bool) {
+	//This mutex is not for safety, but rather for better batching.
+	//It guarantees that only one copier is reading from the channel at one time
+	//This ensures bigger batches as well as less spread of a
+	//single http request to multiple copiers, decreasing latency via less sync
+	//overhead especially on low-pressure systems.
+	getBatchMutex.Lock()
+	defer getBatchMutex.Unlock()
+	req, ok := <-in
```

Comment:
still thinking out loud:  I don't love that we're sleeping while holding the lock though we'd need it for the "hot" gather anyway, and I don't see a better way to get the same performance.

To summarize my understanding:
The lock changes the coordination scheme to something like a token model (sorry I don't know the actual name). One thread has the token, and is allowed to read from the channel. Once that thread fills up (either receives enough messages, or spends enough time) it releases the token, allowing another thread to start receiving, and forwards its messages to the DB. One way of thinking about this is that by default we send batches at 50Hz, but if the batches get too large we start sending faster to pin the batches at `maxCopyRequestsPerTxn`.

Reply:
The "bad" outcome for this is that we'll round-robin across threads in the low-contention setting: after a thread sends a batch it'll release the lock, and, since go mutexen are approximately fair, the woken thread will be the last thread to get a new batch.

One way to fix this would be to add yet another layer of indirection: have this batching be done by a distributor thread that distributes to one or more inserter routines, starting with exactly one routine, and having the distributor spin up more as needed. I not sure this is worth the complexity until we actually see a performance issue.